File: windows/src/Glyphstroke.App/KeyMap.cs
namespace Glyphstroke.App;

/// <summary>
/// Разбор записи вроде <c>ctrl+shift+t</c> в коды виртуальных клавиш Windows.
/// </summary>
/// <remarks>
/// Имена принимаются и в здешнем виде, и в том, что лежит в файлах жестов от
/// Linux и macOS (<c>Page_Up</c>, <c>bracketleft</c>, <c>cmd</c>): файлы у трёх
/// версий общие, и запись, перенесённая с другой машины, должна хотя бы
/// разбираться. А вот сами сочетания перенести нельзя, и подменять их за спиной
/// человека нельзя тем более — для этого есть тип действия «стандартное».
/// </remarks>
public static class KeyMap
{
    private static readonly Dictionary<string, ushort> Keys = new(StringComparer.OrdinalIgnoreCase)
    {
        // модификаторы
        ["ctrl"] = 0x11, ["control"] = 0x11,
        ["alt"] = 0x12, ["option"] = 0x12, ["opt"] = 0x12,
        ["shift"] = 0x10,
        ["win"] = 0x5B, ["super"] = 0x5B, ["meta"] = 0x5B, ["cmd"] = 0x5B, ["command"] = 0x5B,

        // управление
        ["enter"] = 0x0D, ["return"] = 0x0D,
        ["tab"] = 0x09, ["space"] = 0x20,
        ["backspace"] = 0x08,
        ["delete"] = 0x2E, ["del"] = 0x2E, ["forwarddelete"] = 0x2E,
        ["escape"] = 0x1B, ["esc"] = 0x1B,
        ["insert"] = 0x2D, ["ins"] = 0x2D,
        ["home"] = 0x24, ["end"] = 0x23,
        ["pageup"] = 0x21, ["page_up"] = 0x21, ["pgup"] = 0x21, ["prior"] = 0x21,
        ["pagedown"] = 0x22, ["page_down"] = 0x22, ["pgdn"] = 0x22, ["next"] = 0x22,
        ["left"] = 0x25, ["up"] = 0x26, ["right"] = 0x27, ["down"] = 0x28,
        ["printscreen"] = 0x2C, ["print"] = 0x2C, ["prtsc"] = 0x2C,
        ["capslock"] = 0x14, ["menu"] = 0x5D, ["apps"] = 0x5D,

        // знаки: имена в стиле X11 — в перенесённых файлах встречаются именно они
        ["minus"] = 0xBD, ["-"] = 0xBD,
        ["equal"] = 0xBB, ["="] = 0xBB, ["plus"] = 0xBB,
        ["bracketleft"] = 0xDB, ["["] = 0xDB,
        ["bracketright"] = 0xDD, ["]"] = 0xDD,
        ["backslash"] = 0xDC, ["\\"] = 0xDC,
        ["semicolon"] = 0xBA, [";"] = 0xBA,
        ["apostrophe"] = 0xDE, ["'"] = 0xDE,
        ["grave"] = 0xC0, ["`"] = 0xC0,
        ["comma"] = 0xBC, [","] = 0xBC,
        ["period"] = 0xBE, ["."] = 0xBE,
        ["slash"] = 0xBF, ["/"] = 0xBF,
    };

    static KeyMap()
    {
        for (char letter = 'a'; letter <= 'z'; letter++)
        {
            Keys[letter.ToString()] = (ushort)char.ToUpperInvariant(letter);
        }
        for (char digit = '0'; digit <= '9'; digit++)
        {
            Keys[digit.ToString()] = digit;
        }
        for (int number = 1; number <= 24; number++)
        {
            Keys["f" + number] = (ushort)(0x70 + number - 1);
        }
    }

    /// <summary>
    /// <c>ctrl+shift+t</c> → коды клавиш по порядку нажатия; отпускаются они в
    /// обратном.
    /// </summary>
    public static bool TryParse(string combination, out List<ushort> keys)
    {
        keys = new List<ushort>();
        var parts = combination.Split('+', StringSplitOptions.RemoveEmptyEntries);
        if (parts.Length == 0)
        {
            return false;
        }

        var modifiers = new List<ushort>();
        ushort? main = null;
        foreach (var raw in parts)
        {
            string name = raw.Trim();
            if (name.Length == 0)
            {
                continue;
            }
            if (!Keys.TryGetValue(name, out ushort code))
            {
                return false;
            }
            if (IsModifier(code))
            {
                modifiers.Add(code);
            }
            else if (main is null)
            {
                main = code;
            }
            else
            {
                // «ctrl+a+b» — это опечатка, и нажать половину хуже, чем ничего.
                return false;
            }
        }

        if (main is null && modifiers.Count == 0)
        {
            return false;
        }
        keys.AddRange(modifiers);
        if (main is not null)
        {
            keys.Add(main.Value);
        }
        return true;
    }

    private static bool IsModifier(ushort code) => code is 0x11 or 0x12 or 0x10 or 0x5B or 0x5C;
}

File: macos/Sources/GlyphstrokeMac/KeyMap.swift
import Foundation
import CoreGraphics

/// Разбор записи вроде `cmd+shift+t` в код клавиши и набор модификаторов.
///
/// Имена клавиш приняты и в маковском виде (`cmd`, `option`), и в том, что
/// лежит в файлах жестов от Linux (`ctrl`, `Page_Up`, `Return`): файлы у нас
/// общие, и запись, перенесённая с другой машины, должна хотя бы разбираться.
///
/// А вот сами сочетания перенести нельзя, и делать это молча тем более: на
/// Linux «назад» в браузере — `alt+Left`, на маке — `cmd+[`. Подмена `ctrl` на
/// `cmd` за спиной пользователя ломала бы ровно те жесты, где `ctrl` на маке и
/// нужен (в терминале). Поэтому стартовый набор для мака лежит отдельным
/// файлом, а перенесённые жесты человек правит сам.
public enum KeyMap {
    /// Клавиши по именам. Основа — ANSI-раскладка: коды у macOS позиционные,
    /// от языка ввода не зависят.
    public static let keys: [String: CGKeyCode] = [
        "a": 0, "s": 1, "d": 2, "f": 3, "h": 4, "g": 5, "z": 6, "x": 7, "c": 8, "v": 9,
        "b": 11, "q": 12, "w": 13, "e": 14, "r": 15, "y": 16, "t": 17,
        "1": 18, "2": 19, "3": 20, "4": 21, "6": 22, "5": 23,
        "equal": 24, "=": 24, "9": 25, "7": 26, "minus": 27, "-": 27, "8": 28, "0": 29,
        "bracketright": 30, "]": 30, "o": 31, "u": 32, "bracketleft": 33, "[": 33,
        "i": 34, "p": 35, "l": 37, "j": 38, "apostrophe": 39, "'": 39, "k": 40,
        "semicolon": 41, ";": 41, "backslash": 42, "\\": 42, "comma": 43, ",": 43,
        "slash": 44, "/": 44, "n": 45, "m": 46, "period": 47, ".": 47, "grave": 50, "`": 50,

        "return": 36, "enter": 36, "tab": 48, "space": 49,
        "backspace": 51, "delete": 51,            // как на Linux: Delete — это забой
        "forwarddelete": 117, "del": 117,
        "escape": 53, "esc": 53,

        "f1": 122, "f2": 120, "f3": 99, "f4": 118, "f5": 96, "f6": 97, "f7": 98,
        "f8": 100, "f9": 101, "f10": 109, "f11": 103, "f12": 111,
        "f13": 105, "f14": 107, "f15": 113, "f16": 106, "f17": 64, "f18": 79, "f19": 80,

        "home": 115, "end": 119, "page_up": 116, "pageup": 116, "prior": 116,
        "page_down": 121, "pagedown": 121, "next": 121,
        "left": 123, "right": 124, "down": 125, "up": 126,
        "help": 114, "insert": 114,
    ]

    /// Модификаторы: и маковские имена, и те, что встречаются в файлах Linux.
    public static let modifiers: [String: CGEventFlags] = [
        "cmd": .maskCommand, "command": .maskCommand, "super": .maskCommand,
        "meta": .maskCommand, "win": .maskCommand,
        "ctrl": .maskControl, "control": .maskControl,
        "alt": .maskAlternate, "option": .maskAlternate, "opt": .maskAlternate,
        "shift": .maskShift,
        "fn": .maskSecondaryFn,
    ]

    public struct Combination {
        public var key: CGKeyCode
        public var flags: CGEventFlags
    }

    /// `cmd+shift+t` → код клавиши и модификаторы; `nil` — не разобрали.
    public static func parse(_ text: String) -> Combination? {
        let parts = text.split(separator: "+").map {
            $0.trimmingCharacters(in: .whitespaces)
        }.filter { !$0.isEmpty }
        guard !parts.isEmpty else { return nil }

        var flags: CGEventFlags = []
        var key: CGKeyCode?
        for part in parts {
            let name = part.lowercased()
            if let modifier = modifiers[name] {
                flags.insert(modifier)
                continue
            }
            // Клавиша в записи одна: «ctrl+a+b» — это опечатка, и лучше
            // отказаться целиком, чем нажать наугад половину.
            if key != nil { return nil }
            key = keys[name]
            if key == nil { return nil }
        }
        guard let key else { return nil }
        return Combination(key: key, flags: flags)
    }
}
